march_gazebo_plugins/src/com_controller.cpp

```diff
@@ -0,0 +1,210 @@
+    this->model = _parent;
+    this->foot_left = this->model->GetLink("ankle_plate_left");
+    this->foot_right = this->model->GetLink("ankle_plate_right");
+
+    this->swing_step_size = 0.7;  // This estimate will be adjusted every step
+    this->subgait_name = "home_stand";
+
+    this->p_pitch = 60;
+    this->d_pitch = 10;
+    this->p_roll = 150;
+    this->d_roll = 7.5;
+    this->p_yaw = 500;
+    this->d_yaw = 25;
+
+    this->error_x_last_timestep = 0;
+    this->error_y_last_timestep = 0;
+    this->error_yaw_last_timestep = 0;
+
+    // Listen to the update event. This event is broadcast every
+    // simulation iteration.
+    this->updateConnection = event::Events::ConnectWorldUpdateBegin(std::bind(&ComController::OnUpdate, this));
+
+    // Create our ROS node.
+    this->rosNode.reset(new ros::NodeHandle("gazebo_client"));
+
+    // Create a named topic, and subscribe to it.
+    ros::SubscribeOptions so = ros::SubscribeOptions::create<march_shared_resources::GaitActionGoal>(
+        "/march/gait/schedule/goal", 1, boost::bind(&ComController::OnRosMsg, this, _1), ros::VoidPtr(),
+        &this->rosQueue);
+    this->rosSub = this->rosNode->subscribe(so);
+
+    // Spin up the queue helper thread.
+    this->rosQueueThread = std::thread(std::bind(&ComController::QueueThread, this));
+  }
+
+public:
+  void OnRosMsg(const march_shared_resources::GaitActionGoalConstPtr& _msg)
+  {
+    if (this->subgait_name == "right_open" or this->subgait_name == "right_swing" or this->subgait_name == "left_swing")
+    {
+      this->swing_step_size = 0.8 * this->swing_step_size + 0.4 * std::abs(this->foot_right->WorldPose().Pos().X() -
+                                                                           this->foot_left->WorldPose().Pos().X());
+    }
+
+    this->subgait_name = _msg->goal.current_subgait.name;
+    this->subgait_duration =
+        _msg->goal.current_subgait.duration.sec + 0.000000001 * _msg->goal.current_subgait.duration.nsec;
+    this->subgait_start_time = this->model->GetWorld()->SimTime().Double();
+  }
+
+  /// \brief ROS helper function that processes messages
+private:
+  void QueueThread()
+  {
+    static const double timeout = 0.01;
+    while (this->rosNode->ok())
+    {
+      this->rosQueue.callAvailable(ros::WallDuration(timeout));
+    }
+  }
+
+  // Called by the world update start event
+private:
+  double GetMass()
+  {
+    double mass;
+    for (auto const& link : this->model->GetLinks())
+    {
+      mass += link->GetInertial()->Mass();
+    }
+    return mass;
+  }
+
+  // Called by the world update start event
+private:
+  ignition::math::v4::Vector3<double> GetCom()
+  {
+    ignition::math::v4::Vector3<double> com(0.0, 0.0, 0.0);
+    for (auto const& link : this->model->GetLinks())
+    {
+      com += link->WorldCoGPose().Pos() * link->GetInertial()->Mass();
+    }
+    return com / this->GetMass();
+  }
+
+  // Called by the world update start event
+public:
+  void OnUpdate()
```

Comment:
I would move this public method `OnUpdate` up so that the class starts with `public` interfaces and then `private` interfaces.